src/Microsoft.ML.Core/Data/Repository.cs

```diff
@@ -256,7 +256,7 @@ protected void GetPath(out string pathEnt, out string pathTemp, string dir, stri
             string root = Path.GetFullPath(DirTemp ?? @"x:\dummy");
             string entityPath = Path.Combine(root, dir ?? "", name);
             entityPath = Path.GetFullPath(entityPath);
-            string tempPath = Path.Combine(root, PathMap.Count.ToString());
+            string tempPath = Path.Combine(root, Path.GetRandomFileName());
             tempPath = Path.GetFullPath(tempPath);
 
             string parent = Path.GetDirectoryName(entityPath);
```

Comment:
The current function here, `GetPath()`, creates the path. It's called below in `CreateEntry()`:
https://github.com/dotnet/machinelearning/blob/c29a00b5f6089fb5eba17b26f826b791e63236c4/src/Microsoft.ML.Core/Data/Repository.cs#L335-L339

There another minor race condition in how the new path is being checked/added to the dictionary. It's a nit, and minor enough to ignore if you want.

If we want to handle in this PR, we could fix by either adding a lock around the IF/ELSE, or by moving to a [ConcurrentDictionary](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2) and using its `TryAdd()` to decide if to throw.

_(I tagged this on a line of code to allow for a threaded discussion)_

Reply:
Went the `ConcurrentDictionary` route, but let me know if that's not the correct implementation of it.